### PR TITLE
refactored table writing in BasicSomaticValidator

### PR DESCRIPTION
**Pull request by davidbenjamin**
@samuelklee This is for you regarding #3884.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java [hunk 1]
```diff
@@ -104,4 +118,112 @@ public String getFilters() {
     public long getNumAltSupportingReadsInNormal() {
         return numAltSupportingReadsInNormal;
     }
+
+    //----- The following two public static methods read and write files
+    public static void writeToFile(final List<BasicValidationResult> records, final File outputTable) {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java]: I think just `write` and `read` as method names are fine, but up to you.  Might also consider just `file` for the parameter names, or perhaps `outputFile` and `inputFile` (the distinction between the current `outputTable` and `tableFile` is not clear).

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java [hunk 2]
```diff
@@ -104,4 +118,112 @@ public String getFilters() {
...
+        try ( final BasicValidationResult.BasicValidationResultTableWriter writer = new BasicValidationResult.BasicValidationResultTableWriter(IOUtils.fileToPath(outputTable)) ) {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java]: Some inconsistencies in white space here and in the try statement below.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java [hunk 3]
```diff
@@ -104,4 +118,112 @@ public String getFilters() {
...
+            writer.writeHeaderIfApplies();
+            writer.writeAllRecords(records);
+        } catch (IOException e){
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java]: Make the exception final here and below.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java [hunk 4]
```diff
@@ -104,4 +118,112 @@ public String getFilters() {
...
+    public static List<BasicValidationResult> readFromFile(final File tableFile) {
+        try( final BasicValidationResult.BasicValidationResultTableReader reader = new BasicValidationResult.BasicValidationResultTableReader(IOUtils.fileToPath(tableFile)) ) {
+            return reader.toList();
+        } catch (IOException e){
+            throw new UserException(String.format("Encountered an IO exception while reading from %s.", tableFile));
+        }
+    }
+
+    //-------- The following methods are boilerplate for reading and writing tables
+    private static class BasicValidationResultTableWriter extends TableWriter<BasicValidationResult> {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java]: Reconsider the access modifiers of the below classes (and their constructors) and enum (and its `COLUMNS` field)---some inconsistencies here.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java [hunk 5]
```diff
@@ -2,6 +2,20 @@
 
 import htsjdk.samtools.util.Locatable;
 import htsjdk.variant.variantcontext.Allele;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.tools.walkers.contamination.ContaminationRecord;
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java (hunk 1 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java (hunk 2 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java (hunk 3 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java (hunk 4 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/basicshortmutpileup/BasicValidationResult.java]: done

**Pull request merged**